File: java/src/main/java/com/jackson/chapter1/Problem1_1.java
package com.jackson.chapter1;

import java.util.Arrays;

/**
 * Given as array of integers, return a new array such that each element
 * at index i of the new array is the product of all the numbers in the original
 * array except the one at i.
 */
public class Problem1_1 {

    /**
     *
     * @param input array of integers
     * @return array with products of all items except item at i.
     */
    public static int[] createProductArray(int[] input) {
        int[] output = new int[input.length];

        int sum = Arrays.stream(input).reduce((i, last)-> i*last).getAsInt();

        for (int i = 0; i < output.length; i++) {
            output[i] = sum / input[i];
        }
        return output;
    }
}

File: java/src/main/java/com/jackson/chapter1/Problem1_1.kt
package com.jackson.chapter1

fun createProductSum(input: IntArray): IntArray {
    val sum = input.reduce { acc, i -> acc * i }
    return input.map { sum / it }.toIntArray()
}

fun createProductSumWithoutDiv(input: IntArray): IntArray {
    val leftProduct = IntArray(input.size)
    val rightProduct = IntArray(input.size)

    for (i in (0 until input.size)) {
        if (i == 0) {
            leftProduct[i] = 1
        } else {
            leftProduct[i] = leftProduct[i - 1] * input[i - 1]
        }
    }

    for (i in input.size - 1 downTo 0) {
       if (i == input.size - 1) {
           rightProduct[i] = 1
       } else {
           rightProduct[i] = rightProduct[i + 1] * input[i + 1]
       }
    }

    return (0 until input.size).map { leftProduct[it] * rightProduct[it] }.toIntArray()
}